IBM/ibm-common-service-operator
# Create backup_preload_mongo.sh

Create new script for backup and restore of mongo database into a new namespace. The difference between this script and the existing backup_restore_mongo.sh is when the restore takes place. In the original, the restore takes place after the new mongo and IAM instances are ready. This preload script is run before the new mongo is deployed, populating the pvc/pv the new mongo will be using before it is deployed. This avoids some of the problems the backup_restore_mongo.sh experiences, especially on ROKS.

Sequence of events described here https://github.ibm.com/IBMPrivateCloud/roadmap/blob/master/feature-specs/CloudPak3.0/UpgradeSequence.md under `Prepare a MongoDB backup` bullet.

Testing:
- setup environment with multiple cloud paks sharing one cs instance
- create and populate LDAP, verify login to original cs instance
- create target namespace for new cs instance
- run `./backup_preload_mongo.sh <original cs namespace> <target namespace>`
- verify script completes and there is a stand alone `mongodbdir-icp-mongodb-0` in the target namespace
- create csmaps cm with target namespace as a `map-to-namespace` value
- run conversion script
- wait for IAM to come ready
- login to cp-console for new cs instance with an LDAP user created earlier

In my testing this is in a pretty good spot but there are some tweaks left to be made:
- address pod security warnings when creating dummy mongo statefulset
- cs-mongodump pvc can be in "Lost" state in target namespace
    - in my limited experience with this issue, it has not affected the data being transferred

Part of https://github.com/IBM/ibm-common-service-operator/pull/1099 moved to its own PR for clarity.
